### ISSUE TITLE
Make it compatible with cockroachdb

### ISSUE THREAD

**Issue by Meai1** [sfackler/rust-postgres#171]
cockroachdb is compatible with the postgres wire protocol but not with all the postgres internals. Many other postgres drivers work fine with cockroachdb, but this one doesnt!

Error:

> thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Io(Error { repr: Custom(Custom { kind: InvalidInput, error: StringError("the server returned an unexpected response") }) })', ../src/libcore/result.rs:688

Repro:

```
wget https://binaries.cockroachdb.com/cockroach-beta-20160407.linux-amd64.tgz && tar xfz cockroach-beta-20160407.linux-amd64.tgz && cd cockroach-beta-20160407.linux-amd64
./cockroach start
./cockroach sql 
> CREATE DATABASE root;
```

 ```
 let conn = Connection::connect("postgresql://root@localhost:26257/root?sslmode=disable", SslMode::None) .unwrap();
	println!("connection established");
    conn.execute("CREATE TABLE accounts (id INT PRIMARY KEY, balance INT)", &[]).unwrap();
	println!("executed statement");
```

**Comment by sfackler** [comment]
Cockroach doesn't appear to send back the file, line, and routine fields in its error responses, which Postgres will always return: http://www.postgresql.org/docs/current/static/protocol-error-fields.html.

In addition, the sqlstate error code it returns is "XX000", corresponding to "internal error", rather than the "42P01" ("undefined table") that Postgres will return.

The error is being thrown because Cockroach doesn't emulate any of the pg_catalog tables. This can be hacked around, but not in a particularly reasonable way until the error returns are made consistent with Postgres.

**Comment by Meai1** [comment]
Are you sure this is a cockroach problem? Because I tried to connect to it with the dart postgresql driver and it returned errors and info just fine, everything worked.
If you look at the list of supported drivers on their website, I doubt they all made custom modifications to their sources to support cockroach: https://www.cockroachlabs.com/docs/install-client-drivers.html

Anyway, if there is no other way: Can you tell me what I have to do to make this work, even if it's hacky?

**Comment by sfackler** [comment]
It sounds like Dart's driver doesn't depend on the features that Cockroach's interface doesn't yet support.

This branch has the changes needed to make that test case work: https://github.com/sfackler/rust-postgres/tree/cockroach-hacks. I don't feel super comfortable merging these into the mainline branch. I'll file an issue with Cockroach about tweaking the error response.

**Comment by sfackler** [comment]
This looks to be the issue to follow for the bad error code: https://github.com/cockroachdb/cockroach/issues/5329

**Comment by sfackler** [comment]
It looks like pg_catalog is not a planned addition: https://github.com/cockroachdb/cockroach/issues/5194

**Comment by Meai1** [comment]
yeah i read about that part and that's why I uncommented the whole "setup_typeinfo_query" function but it wasnt enough (didnt change anything) for me so that's why I opened this issue

**Comment by sfackler** [comment]
That sounds reasonable I think - I don't expect people are looking at those fields very often anyway! I'll make the change to use `Option`s in the next breaking release, but in the meantime, I think it's probably okay to just paper over the issue by filling in empty strings and 0s where necessary: https://github.com/sfackler/rust-postgres/blob/6567e3e56458f24421788d7e0e3b647b3251943a/src/error/mod.rs#L116-L118.

With respect to `pg_catalog`, it seems fine to skip the typeinfo query setup steps if it's missing provided Cockroach doesn't use any types that aren't built into Postgres, which I think is the case? However, I do feel a bit leery about accepting InternalError vs UndefinedTable: https://github.com/sfackler/rust-postgres/blob/6567e3e56458f24421788d7e0e3b647b3251943a/src/lib.rs#L424-L426, so corrected error codes would definitely be appreciated 😄 .

**Comment by nvanbenschoten** [comment]
Thanks, that would be really helpful! In the meantime, I'll look into either mocking out the error fields or fully supporting them.

All of our data types map directly to Postgres Object IDs, so if skipping the typeinfo query setup steps is possible when not supported, that would be a great first step. While we don't have plans to directly support `pg_catalog` (because it's explicitly Postgres specific), there are plans to support `information_schema`, which looks like a close parallel that is actually part of the SQL standard.

With regards to the error codes, is UndefinedTable the only error that you need support for on our end? We're slowly [adding support](https://github.com/cockroachdb/cockroach/blob/master/sql/errors.go#L31) for Postgres error codes as we come upon them (it's hard to know which ones to prioritize out of [all of these](http://www.postgresql.org/docs/9.5/static/errcodes-appendix.html)), so if that's the case I can look into adding support.

**Comment by sfackler** [comment]
> All of our data types map directly to Postgres Object IDs, so if skipping the typeinfo query setup steps is possible when not supported, that would be a great first step. While we don't have plans to directly support pg_catalog (because it's explicitly Postgres specific), there are plans to support information_schema, which looks like a close parallel that is actually part of the SQL standard.

Sounds good - I don't foresee rust-postgres needing to poke at any metadata tables other than what we do now for type lookup, so a lack of pg_catalog should be fine.

UndefinedTable is the only code that rust-postgres looks at programmatically.

I made the change to the line/file/routine fields on the `breaks` branch and it'll land in the next breaking release: https://github.com/sfackler/rust-postgres/commit/3df519ee38484f9c94fbf293c4d1bee1e31a8fb3

**Comment by nvanbenschoten** [comment]
I just merged in both of the changes we discussed above, but I'm still running into issues with the pg_catalog queries. Specifically, you mentioned that 

> UndefinedTable is the only code that rust-postgres looks at programmatically.

Unfortunately, but not completely surprisingly, when querying for tables like `pg_catalog.pg_type`, CockroachDB is actually returning an undefined database error (`3D000 - InvalidCatalogName`), because the entire `pg_catalog` database is missing. It sounds like a lack of a pg_catalog database should be fine, so I'm hoping you would be open to merging something along the lines of [this](https://github.com/nvanbenschoten/rust-postgres/commit/ab2f3305e1a14c63ed25a7f6ef0d9ae2b6187c48). With that last change, I've been able to get a proof of concept demo app working!

**Comment by sfackler** [comment]
Postgres itself appears to return `42P01` `UndefinedTable` rather than `InvalidCatalogName` for these kinds of cases for whatever reason. If it'd be easy on your end to change that it'd be great, but looking for `InvalidCatalogName` seems totally reasonable as well.

**Comment by nvanbenschoten** [comment]
Ah thanks for catching that! It's somewhat interesting behavior from Postgres, given that they return an `InvalidCatalogName` error when attempting to connect to a database that does not exist, but an `UndefinedTable` error when attempting to query a table in a database that does not exist. Because of how our error handling is structured, we don't make a distinction here, and it seems to make a lot more sense to return an `InvalidCatalogName`, as the root cause of the issue is a missing database and not a missing table.

**Comment by sfackler** [comment]
So I believe in SQL, a catalog is more akin to what you'd think of as a database - a collection of schemas: http://stackoverflow.com/a/7023159. `pg_catalog` is specifically not a catalog but rather a schema - the name derives from "postgres-specific information about the current catalog" rather than "the catalog with postgres-specific information".

**Comment by nvanbenschoten** [comment]
Huh, that's pretty interesting. I didn't realize the two had such subtle differences. So it sounds like Postgres is special casing their handling of these "meta-schemas". In that case, I think it makes even more sense to have a change like https://github.com/sfackler/rust-postgres/pull/178, as I'm not sure that it can be expected that all databases which speak Postgres' wire protocol will also handle this special case.

**Comment by sfackler** [comment]
I'm not sure what the special case is here? `UndefinedTable` is what postgres returns when you try to access an invalid table from any schema, including `pg_catalog`:

```
projects=# \set VERBOSITY verbose
projects=# select * from pg_catalog.foo;
ERROR:  42P01: relation "pg_catalog.foo" does not exist
LINE 1: select * from pg_catalog.foo
                      ^
LOCATION:  parserOpenTable, parse_relation.c:1138
projects=# select * from foobar.foo;
ERROR:  42P01: relation "foobar.foo" does not exist
LINE 1: select * from foobar.foo;
                      ^
LOCATION:  parserOpenTable, parse_relation.c:1138
projects=# select * from public.foo;
ERROR:  42P01: relation "public.foo" does not exist
LINE 1: select * from public.foo;
                      ^
LOCATION:  parserOpenTable, parse_relation.c:1138
```

**Comment by nvanbenschoten** [comment]
I guess special case may have been the wrong term there, especially now that I'm looking at the diagram in the second answer. What I was trying to get at is that the added match arm will allow the driver to work well with databases that do not have a distinct schema (namespace) layer. For instance, neither MySQL nor CockroachDB made the design decision to structure schemas and databases at different hierarchical layers: http://stackoverflow.com/questions/7942520/relationship-between-catalog-schema-user-and-database-instance. In those cases, it's reasonable to expect them to interpret the `pg_catalog` part of the relation as a catalog (database) name, and therefore return an `InvalidCatalogName` error.

**Comment by sfackler** [comment]
IMO, if you're a server pretending to be Postgres, you should try to match what Postgres does :smile:. But in any case, `InvalidCatalogName` is a fine thing to check for.

**Comment by nvanbenschoten** [comment]
Awesome, thanks for all the help Steven! At this point, rust-postgres is now compatible with CockroachDB. I'll be making a few changes over the next few days to reflect sample usage of CockroachDB in a Rust test application within our documentation, as well as pointing to this driver
as the recommended client driver for Rust.

**Comment by sfackler** [comment]
Yay! Did we ever resolve what to do with the location fields in errors? I have them slated to become optional in the next major release, but I don't think the temporary fallback support is there right now if they're omitted.

**Comment by nvanbenschoten** [comment]
Yep, I added in support for them in https://github.com/cockroachdb/cockroach/pull/6654, so there isn't any requirement for them to be optional on our end anymore. Speaking of releases though, what are your plans for releasing the last few minor changes?

**Comment by sfackler** [comment]
Ah, great. I'll leave the pending change in - seems like it can't hurt.

Published 0.11.8!
